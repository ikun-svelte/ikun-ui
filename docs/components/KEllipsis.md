---
title: KEllipsis
lang: en-US
---

# KEllipsis

It can omit your text to a certain extent.

## Install

::: code-group

```bash [pnpm]
pnpm add @ikun-ui/ellipsis
```

```bash [yarn]
yarn add @ikun-ui/ellipsis
```

```bash [npm]
npm install @ikun-ui/ellipsis
```

:::

## Basic usage

Basic single line ellipsis with tooltip.  
`content` is the text you need to display in the tooltip, `text` is the text you need to abbreviate.

<demo src="ellipsis/basic.svelte"  github='Ellipsis'></demo>

## Omit the number of characters

You can specify the number of words you need to omit through `num`.

<demo src="ellipsis/num.svelte" github='Ellipsis'></demo>

## omitted position

`placement` can specify the position where the ellipsis appears

<demo src="ellipsis/position.svelte" github='Ellipsis'></demo>

## Omit multiple lines

See [caniuse <span class="i-carbon-link text-12px" />](https://caniuse.com/?search=line-clamp) for compatibility.

<demo src="ellipsis/line-clamp.svelte" github='Ellipsis'></demo>

## Ellipsis Props

| Name      | Type                            | Default | Description                                                    |
| --------- | ------------------------------- | ------- | -------------------------------------------------------------- |
| content   | `string`                        | `''`    | The text content displayed by tooltip.                         |
| text      | `string`                        | `''`    | Text content that needs to be omitted.                         |
| expand    | `boolean`                       | `false` | Abbreviated content trigger event to expand to the full text.. |
| lineClamp | `number \| null`                | `null`  | Maximum lines.                                                 |
| num       | `number`                        | `5`     | Number of characters to omit.                                  |
| placement | `'left' \| 'right' \| 'center'` | `right` | Ellipsis display position.                                     |
| cls       | `string`                        | `''`    | Additional class for component                                 |
| attrs     | `Record<string, string>`        | `{}`    | Additional attributes                                          |
