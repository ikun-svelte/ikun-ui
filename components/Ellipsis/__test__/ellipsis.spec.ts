import { afterEach, expect, test, describe, beforeEach, vi } from 'vitest';
import KEllipsis from '../src';
import { tick } from 'svelte';

let host: HTMLElement;

const initHost = () => {
	host = document.createElement('div');
	host.setAttribute('id', 'host');
	document.body.appendChild(host);
};
beforeEach(() => {
	initHost();
	vi.useFakeTimers();
});
afterEach(() => {
	host.remove();
	vi.restoreAllMocks();
});

describe('Test: KEllipsis', () => {
	test('props: cls', async () => {
		const instance = new KEllipsis({
			target: host,
			props: {
				cls: 'k-ellipsis--test'
			}
		});
		expect(instance).toBeTruthy();
		expect((host as HTMLElement)!.innerHTML.includes('k-ellipsis--test')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: content', async () => {
		const instance = new KEllipsis({
			target: host,
			props: {
				content: '自古逢秋悲寂寥',
				text: '我言秋日胜春朝',
				num: 0
			}
		});
		expect(instance).toBeTruthy();
		const elm = host.querySelector('[slot="triggerEl"]');
		elm!.dispatchEvent(new Event('mouseenter', { bubbles: true }));
		await tick();
		await vi.advanceTimersByTimeAsync(300);
		expect(document.body.innerHTML.includes('我言秋日胜春朝')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: text', async () => {
		const instance = new KEllipsis({
			target: host,
			props: {
				content: '自古逢秋悲寂寥',
				text: '我言秋日胜春朝',
				num: 0
			}
		});
		expect(instance).toBeTruthy();
		expect((host as HTMLElement)!.innerHTML.includes('我言秋日胜春朝')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: num', async () => {
		const instance = new KEllipsis({
			target: host,
			props: {
				content: '自古逢秋悲寂寥',
				text: '我言秋日胜春朝',
				num: 3
			}
		});
		expect(instance).toBeTruthy();
		expect((host as HTMLElement)!.innerHTML.includes('我言秋日...')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: expand', async () => {
		const instance = new KEllipsis({
			target: host,
			props: {
				content: '自古逢秋悲寂寥',
				text: '我言秋日胜春朝',
				num: 3,
				expand: true,
				lineClamp: 3
			}
		});
		expect(instance).toBeTruthy();
		const elm = host.querySelector('[slot="triggerEl"]') as HTMLElement;
		expect(elm.style.display === '-webkit-inline-box').toBeTruthy();
		elm.dispatchEvent(new Event('click', { bubbles: true }));
		await tick();
		expect(elm.style.display === '').toBeTruthy();
		elm.dispatchEvent(new Event('click', { bubbles: true }));
		await tick();
		expect(elm.style.display === '-webkit-inline-box').toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: placement', async () => {
		const instance = new KEllipsis({
			target: host,
			props: {
				content: '自古逢秋悲寂寥',
				text: '我言秋日胜春朝',
				placement: 'center',
				num: 3
			}
		});
		expect(instance).toBeTruthy();
		expect((host as HTMLElement)!.innerHTML.includes('我言...春朝')).toBeTruthy();

		instance.$set({ placement: 'left' });
		await tick();
		expect((host as HTMLElement)!.innerHTML.includes('...日胜春朝')).toBeTruthy();

		instance.$set({ placement: 'right' });
		await tick();
		expect((host as HTMLElement)!.innerHTML.includes('我言秋日...')).toBeTruthy();
	});
});
