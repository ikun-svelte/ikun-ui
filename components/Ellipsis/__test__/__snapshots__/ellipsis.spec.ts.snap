// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test: KEllipsis > props: cls 1`] = `"<div><div aria-hidden=\\"true\\"><div class=\\"k-ellipsis k-ellipsis--base k-ellipsis--test\\" slot=\\"triggerEl\\" aria-hidden=\\"true\\" style=\\"\\">...</div></div> </div>"`;

exports[`Test: KEllipsis > props: content 1`] = `"<div><div aria-hidden=\\"true\\"><div class=\\"k-ellipsis k-ellipsis--base\\" slot=\\"triggerEl\\" aria-hidden=\\"true\\" style=\\"\\">我言秋日胜春朝</div></div> <div class=\\"k-popover--base k-popover--base__top k-popover--base__top__dark k-popover--base__dark svelte-13arsd0\\" data-popper-placement=\\"top\\" aria-hidden=\\"true\\" style=\\"position: fixed; left: 0px; top: 0px; margin: 0px; animation: __svelte_3896445108_0 200ms linear 0ms 1 both; bottom: 8px;\\" data-popper-reference-hidden=\\"\\" data-popper-escaped=\\"\\"><span slot=\\"contentEl\\">自古逢秋悲寂寥</span> <div k-popover-arrow=\\"true\\" class=\\"svelte-13arsd0\\" data-popper-arrow-top=\\"\\"></div></div></div>"`;

exports[`Test: KEllipsis > props: expand 1`] = `"<div><div aria-hidden=\\"true\\"><div class=\\"k-ellipsis k-ellipsis--base\\" slot=\\"triggerEl\\" aria-hidden=\\"true\\" style=\\"-webkit-line-clamp: 3; display: -webkit-inline-box; overflow: hidden;\\">我言秋日...</div></div> </div>"`;

exports[`Test: KEllipsis > props: num 1`] = `"<div><div aria-hidden=\\"true\\"><div class=\\"k-ellipsis k-ellipsis--base\\" slot=\\"triggerEl\\" aria-hidden=\\"true\\" style=\\"\\">我言秋日...</div></div> </div>"`;

exports[`Test: KEllipsis > props: text 1`] = `"<div><div aria-hidden=\\"true\\"><div class=\\"k-ellipsis k-ellipsis--base\\" slot=\\"triggerEl\\" aria-hidden=\\"true\\" style=\\"\\">我言秋日胜春朝</div></div> </div>"`;
